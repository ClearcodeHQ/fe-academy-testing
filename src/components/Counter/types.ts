export interface CounterProps {
  onIncrement?: () => void;
  onDecrement?: () => void;
}
