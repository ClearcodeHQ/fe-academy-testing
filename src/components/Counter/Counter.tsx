import React from 'react';
import { useState } from 'react';
import { CounterProps } from './types';

export const Counter = ({ onIncrement, onDecrement }: CounterProps) => {
  const [count, setCount] = useState(0);

  const increment = () => {
    if (onIncrement) {
      onIncrement();
    }
    setCount((prevState) => prevState + 1);
  };

  const decrement = () => {
    if (onDecrement) {
      onDecrement();
    }
    setCount((prevState) => prevState - 1);
  };

  return (
    <>
      <span>Counter</span>
      <button onClick={increment}>increment</button>
      <button onClick={decrement}>decrement</button>
      <span>count: {count}</span>
    </>
  );
};
