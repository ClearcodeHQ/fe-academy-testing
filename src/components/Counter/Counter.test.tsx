import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import { Counter } from './Counter';

describe('Counter', () => {
    it('should match the snapshot', () => {
      const { asFragment } = render(<Counter />);

      expect(asFragment()).toMatchSnapshot();
    });

    it('should increment the count', async () => {
    render(<Counter />);

    const button = screen.getByRole('button', { name: /increment/ });

    const count = screen.getByText(/count/);

    expect(count).toHaveTextContent('count: 0');

    await userEvent.click(button);

    expect(count).toHaveTextContent('count: 1');
  });

  it('should call onIncrement function on counter increment', async () => {
    const onIncrement = jest.fn();

    render(<Counter onIncrement={onIncrement} />);

    const button = screen.getByRole('button', { name: /increment/ });

    await userEvent.click(button);

    expect(onIncrement).toBeCalledTimes(1);
  });

  it('should decrement the count', async () => {
    render(<Counter />);

    const button = screen.getByRole('button', { name: /decrement/ });

    const count = screen.getByText(/count/);

    expect(count).toHaveTextContent('count: 0');

    await userEvent.click(button);

    expect(count).toHaveTextContent('count: -1');
  });

  it('should call onDecrement function on counter decrement', async () => {
    const onDecrement = jest.fn();

    render(<Counter onDecrement={onDecrement} />);

    const button = screen.getByRole('button', { name: /decrement/ });

    await userEvent.click(button);

    expect(onDecrement).toBeCalledTimes(1);
  });
});
