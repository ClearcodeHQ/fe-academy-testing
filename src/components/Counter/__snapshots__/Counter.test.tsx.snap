// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter should match the snapshot 1`] = `
<DocumentFragment>
  <span>
    Counter
  </span>
  <button>
    increment
  </button>
  <button>
    decrement
  </button>
  <span>
    count: 0
  </span>
</DocumentFragment>
`;
